web/pdf_history.js

```diff
@@ -376,6 +386,19 @@ class PDFHistory {
       return;
     }
 
+    if (POSITION_UPDATED_THRESHOLD > 0 && this._isPagesLoaded &&
+        this._destination && !this._destination.page) {
+      // If the current destination was set through the user changing the hash
+      // of the document, we will usually not try to push the current position
+      // to the browser history; see `this._tryPushCurrentPosition()`.
+      //
+      // To prevent `this._tryPushCurrentPosition()` from effectively being
+      // reduced to a no-op in this case, we will assume that the position
+      // *did* in fact change if the 'updateviewarea' event was dispatched
+      // more than `POSITION_UPDATED_THRESHOLD` times.
```

Comment:
I'm not entirely sure what this means or what this commit aims to achieve. Why is it bad if `_tryPushCurrentPosition` is reduced to a no-op if the hash is invalid, and where does the heuristic come from? Could you elaborate a bit on what problem this commit solves and how it works (approximately)?

Reply:
Consider loading the following document http://ftp.acc.umu.se/mirror/CTAN/info/lshort/english/lshort.pdf#pagemode=thumbs
In this case, the hash is defined, but it will obviously not cause a destination to be navigated to (which would set the `page` of its browser history entry). Furthermore, `parseCurrentHash` can obviously not extract a `page` either; https://github.com/mozilla/pdf.js/pull/8775#discussion_r135468381 might be somewhat relevant here too.

Given https://github.com/mozilla/pdf.js/pull/8775/commits/1f41a595995baedfd9265ad93e226ba0a5a89c0c#diff-3b4ebbe223f93c9123a754b4845844e7R288, `tryPushCurrentPosition` won't ever update the browser history with the current position (unless `push` is called). This could thus mean that despite commit https://github.com/mozilla/pdf.js/pull/8775/commits/281070b20c3ed17979b85482922d9245c8eee623, the browser history might not accurately capture the current position when the viewer is navigated away from or the browser closed.

Hence this heuristic, that attempts to "unlock" the adding of entries to the browser history when the position has updated enough times that it should be safe to assume that the position has in fact changed.
You could argue that this is a bit magical, especially since the value of the constant was basically pulled out of a hat, but it's my attempt to avoid permanently disabling certain parts of the implementation in e.g. the edge-case outlined above.
Note that in most cases, this heuristic won't even be necessary, but I figured that it shouldn't hurt to try and avoid this situation.

Does this help, if not ask again!